qadence/ml_tools/train_utils/strategy.py

```
@@ -0,0 +1,350 @@
+            backend (str): Backend to use for distributed communication (default: "nccl").
+        """
+        self.backend: str = backend
+        self.compute_setup: str = compute_setup
+        self.log_setup: str = log_setup
+        self.strategy: str
+        self.rank: int
+        self.world_size: int
+        self.local_rank: int | None
+        self.master_addr: str
+        self.master_port: str
+        self.compute: str
+        self.device: str
+        self.log_device: str
+        self.spawn: bool
+        self.nprocs: int
+        self.dtype: torch.dtype | None = dtype
+        self.data_dtype: torch.dtype | None = None
+        # TODO: This is legacy behavior of data_dtype. Modify it appropriately.
+        if self.dtype:
+            self.data_dtype = torch.float64 if (self.dtype == torch.complex128) else torch.float32
+        self._set_cluster_variables()
+
+    def detect_strategy(self) -> str:
+        """
+        Detect the launch strategy based on environment variables.
+
+        Returns:
+            str: The detected launch strategy. Possible values are:
+                - "Default": If the "LOCAL_RANK" environment variable is set.
+                    Possibly by torchrun or my spawned processes.
+                - "slurm": If the "SLURM_PROCID" environment variable is set.
+                - "none": Otherwise.
+        """
+        if ("LOCAL_RANK" in os.environ) or self.spawn:
+            return "default"
+        elif "TORCHELASTIC_RUN_ID" in os.environ:
+            return "torchrun"
+        else:
+            return "none"
+
+    def _set_cluster_variables(self) -> None:
+        """
+        Sets the initial default variables for the cluster.
+
+        For now it only supports SLURM Cluster, and should be extended to others
+        when needed.
+        """
+        self.job_id = int(os.environ.get("SLURM_JOB_ID", 93345))
```

Why 93345 as default?

---

Just a random job ID I was testing it on. Can also change to a random.choice - but that might lead to issues in identifying ports - as unavailable ports might get selected.

---

self.job_id is an indicator used to assign id to the execution (and to create a unique port every run). It has nothing to do with slurm. It should be assigned for every run, on local run/slurm cluster/any other cluster. 

SLURM_JOB_ID is being used as a proxy for this self.job_id - as its always unique - and we wont face port unavailable issues on the cluster. 


On your local run - the localhost has port 21069 (based on default job id 93345 ((93345 + 12000) / 5000))
On slurm cluster - the nodexx has port xxxxx (based on the slurm job id ((SLURM_JOB_ID + 12000) / 5000))

I don't think we should raise an error. What we can do is change the default value, which will also change the default port. We can either hard-code it, or, randomly generate the job_id (where we risk getting port unavailability.)

---

Alright - I have updated it so that we decouple job_id and port. 
Port is identified using a _find_avaliable_port function. 
And job_id is "Unknown" as a default.